### PR TITLE
[WIP] Rust: 1.9.0 -> 1.10.0

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

Besides the version update, I restructured things to match GHC, which I think sets a good precedent for us. Need to get hashes for beta, and actually test building.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ericson2314** [comment]
CC @solson, who wanted a ping, and @wycats who might find this interesting.

**Comment by dvc94ch** [comment]
@Ericson2314: I know it's confusing, but head is meant to be built by stable, this is not a mistake - even dough it says beta in stage0.txt

**Comment by dvc94ch** [comment]
pinging @the-kenny because he's probably interested too and moves thing along... :)

some other cool stuff to be done is using cargo to actually build rustc and putting cargo packages into separate store paths instead of grouping them all into one. I won't be getting round to this since I'm currently working on other awesome projects, there so many of them!!

**Comment by Ericson2314** [comment]
@dvc94ch good job with the previous refactor by the way. This has gotten a lot better since I last touched it a year or two ago!

I'm confused by the link---I assume you were mentioning the "(or stable in the future, probably after the next release)" parenthetical? I asked on IRC and was told that everything is indeed built from the last release, and for random master commits / nightly, the "last release" is the last beta.   

> some other cool stuff to be done is using cargo to actually build rustc and putting cargo packages into separate store paths instead of grouping them all into one. I won't be getting round to this since I'm currently working on other awesome projects, there so many of them!!

Oh great! do you have a link? I am very excited for these sorts of things.

**Comment by Ericson2314** [comment]
How does one find the `depsSha256`?

**Comment by dvc94ch** [comment]
If you look at an example package in the registry [0] there is a checksum. It should be possible to harvest the sha256 hash from there and create a derivation manually (instead of using the src=fetchurl sugar). 

Currently cargo build/fetch is used to fetch the dependencies and they are put into $out by setting `CARGO_HOME` [1]. I think it might be possible to loop through each folder in $CARGO_HOME/target/release, compute the store location - if store location available skip else create a derivation and send it to nix-daemon. At this point the dependency is already compiled so it would just copy the binary into a store location.

Then I don't know exactly how rustc/cargo find binaries, but there should be some environment variable like LIBRARY_PATH or LD_LIBRARY_PATH you can set.

This would be my (initial) approach, maybe you have a better idea?

[0] https://github.com/rust-lang/crates.io-index/blob/master/2/an
[1] https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/fetch-cargo-deps#L18

**Comment by Ericson2314** [comment]
@dvc94ch 

Ah, I didn't realize that Cargo is on crates.io, but since it is, all is good!

Regarding the longer term things, https://github.com/rust-lang/rfcs/pull/1133 is my (soon to be rewritten) grand plan for making Cargo (optionally) not rely on rustc lib finding hacks. One step closer to simplicity.

**Comment by dvc94ch** [comment]
Sorry if my half-backed brainstorming wasn't useful to you.

**Comment by Ericson2314** [comment]
@dvc94ch no not at all, it was! I just assumed it was downloading from source, so I'd have to hash lockfile or something myself. And even if I had known it was on crates.io I didn't know what to do about it wrt the index, so thanks on both counts.

**Comment by Ericson2314** [comment]
Ok, dep hash is there. Time to see what breaks next :)

**Comment by the-kenny** [comment]
FIY: I tried to package 1.10 too (with less changes around our infrastructure) and ran into this upstream bug: https://github.com/rust-lang/rust/issues/34722

**Comment by the-kenny** [comment]
I'm not sure if we should change our rust packaging strategy *again*. Is there a specific reason for this? What are the benefits / drawbacks to our previous strategy?

File: pkgs/top-level/rust-packages.nix [hunk 1]
```diff
@@ -1,49 +1,35 @@
...
+      inherit rust;
+
+      rustRegistry = callPackage ../development/rust-modules { };
+
+      buildRustPackage = callPackage ../build-support/rust {
+        inherit rust;
+      };
+    });
+
+  tools = {
```
Comment (the-kenny) [inline pkgs/top-level/rust-packages.nix]: `tools` doesn't sound like a good name to me.

File: pkgs/top-level/rust-packages.nix [hunk 2]
```diff
@@ -1,49 +1,35 @@
...
+    rust-1_9_0-binary = callPackage ../development/compilers/rust/1.9.0-binary.nix {};
+    rust-1_10_0 = callPackage ../development/compilers/rust/1.10.0.nix {
+      rustPlatform = platforms.rustPlatform-1_9_0-binary;
```
Comment (the-kenny) [inline pkgs/top-level/rust-packages.nix]: Is there a consensus to add `major.minor.patch` to all attribute-paths? I strongly prefer `beta`, `stable`, etc. by default (with an exception for `1.9` for bootstrapping here)

Comment (the-kenny) [inline pkgs/top-level/rust-packages.nix]: Especially I don't think we should keep old versions around (except for bootstrapping, like `1.9.0-binary` here), so explicit versioning in the attribute-paths is quite pointless. Plus every attribute path with a version in it is a place that needs to be updated when a new version comes along, making updates more complicated.

**Comment by the-kenny** [comment]
Hashes are still broken: 

```
output path ‘/nix/store/k4klsy1hwc4dc1cgiwz9hglb6pri7h39-rust’ has r:sha256 hash ‘15i81ybh32xymmkyz3bkb5bdgi9hx8nb0sh00ac6qba6w8ljpii9’ when ‘0xcxnds8ak4yvdk0nzkl35skvagyhc6pgyb45rqmfmp3iswra8p5’ was expected

```

**Comment by Ericson2314** [comment]
@the-kenny all this was motivated by the understanding that `1.9.0` as the last prebuilt binary we'd use, and we'd bootstrap from source for there.

[Long term https://github.com/thepowersgang/mrustc existsly mainly to be able to bootstrap without any pre-built binaries.]

**Comment by dvc94ch** [comment]
That's not accurate... We still need to bootstrap from the release tarball.
While we could in theory use old versions from the binary cache for this,
this makes it really hard if you want to bootstrap nixos for a new platform
or want to opt out of the binary cache and build it yourself. (You have to
build each version consecutively until the latest version, and you have to
keep old versions around to be able to follow the bootstrapping chain). The
only thing that's changed is that we aren't using snapshots anymore but
releases.

**Comment by Ericson2314** [comment]
@dvc94ch I don't get the "keep" old versions around part. You can rebuild at any time though note this takes longer. Also note that `m.n+1` must be built with m.n`.

Most distros (nixpkgs included AFIAK) don't like to rely on to many prebuilt binaries, and if they do, they rather avoid changing prebuilt binary too often.

**Comment by dvc94ch** [comment]
That's because you can build gcc with any other version of gcc (within
bounds) Rust is a lot more restrictive in this regard currently

**Comment by Ericson2314** [comment]
@dvc94ch Haskell has the same restriction and has been bootstrapped in this way for some time.

**Comment by Ericson2314** [comment]
@dvc94ch Ok I'll admit this isn't as deep a chain as I thought but still two hops when only one is needed https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/haskell-packages.nix

**Comment by dvc94ch** [comment]
So in three release windows time(6*3 weeks), if someone wants to build rust
on nixos arm they'll have to download the 1.9.0 binary to compile 1.10 1.11
1.12 1.13? What if for some reason an old version has an arm related bug
and they're stuck in the middle of the bootstrapping chain, submitting
patches for an old version?
On Jul 10, 2016 6:59 PM, "John Ericson" <notifications@github.com> wrote:

@dvc94ch <https://github.com/dvc94ch> Haskell has the same restriction and
has been bootstrapped in this way for some time.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/NixOS/nixpkgs/pull/16807#issuecomment-231599075>, or mute
the thread
<https://github.com/notifications/unsubscribe/AAtRr2GMvnXOEe9FllGxaBpIGoIvRPrXks5qUSTzgaJpZM4JIlL3>
.

**Comment by the-kenny** [comment]
I don't think we should keep a chain around. On a new release we should just integrate the previous `stable` as a `-binary` for bootstrap.

**Comment by dvc94ch** [comment]
That's my opinion too

**Comment by dvc94ch** [comment]
The even bigger issue is if someone makes a change to the rustc package,
all previous versions need to be rebuilt

**Comment by Ericson2314** [comment]
@dvc94ch I'd say break this rule in the event such a problem occurs, only for the affected architecture.

----

Well, if I am the only one that ones this, then clearly no consensus. What do you two think of my first commit (a smaller refactor)?

**Comment by dvc94ch** [comment]
I'm afk, I'll have to get back to you on that one tomorrow.

**Comment by Ericson2314** [comment]
@dvc94ch ok. @the-kenny feel free to open another PR, lest I just waist time chasing down the hashes you already found.

**Comment by the-kenny** [comment]
@Ericson2314 Is there anything wrong? I didn't want to sound too harsh with my comments on your PR.

**Comment by Ericson2314** [comment]
@the-kenny no, not at all. I mean regardless of whether we bootstrap from source or not, I'd need to chase down the hashes---and I personally seem to have very bad luck with the prefetch scripts so usually just wait for build failures. But if you got as far as the secret key stuff, then you must have all the correct hashes.

**Comment by dvc94ch** [comment]
If you use the nix-build --check flag I think you should get the new source
hash...

**Comment by dvc94ch** [comment]
Or change the hash a little so that it doesn't use the cache. That def works

**Comment by Ericson2314** [comment]
@dvc94ch the latter is what I normally do.

**Comment by Ericson2314** [comment]
Actually haven't tried --check yet, I'll look at that.

**Comment by Ericson2314** [comment]
@dvc94ch so the current failure has an expected hash that doesn't turn up if I grep in nixpkgs. How weird is that?!

**Comment by Ericson2314** [comment]
@the-kenny I g2g now, but if you do make a PR with what you have, I'll be glad to assist with the secret key stuff. I can just rename/rebase this to discuss refactors separately. For now the big long bootstrap chains and current way coincide, and I'll probably scale this back to just move the bootstrapping out of all-packages.nix but ditch explicit version names and source-for-ever approach anways :).

**Comment by dvc94ch** [comment]
I think the hashes are a function of the build inputs so the hash of the
source package is probably only part of what makes the store location, but
don't know exactly how it's computed. Would definitely be interested if
someone has some more insight, I haven't gotten round to reading edolstras
doctor thesis yet.

**Comment by the-kenny** [comment]
@Ericson2314 you can find my code at https://github.com/the-kenny/nixpkgs/tree/rust-1.10.0

```
nix-build $path-to-clone -A rustc
``` to reproduce the secret-key issue detailed in https://github.com/rust-lang/rust/issues/34722

**Comment by Ericson2314** [comment]
Thanks!

**Comment by dvc94ch** [comment]
@Ericson2314 Did you actually get 1.10 to build?
@the-kenny I checked out your git branch and it looks to me like you did everything right...

**Comment by Ericson2314** [comment]
Sorry I've been doing other things. I didn't see anything wrong either, maybe we should try rustbuild (the new buildsystem that will eventually replace the make-based one). I'd love to move there anyways.

**Comment by dvc94ch** [comment]
@Ericson2314 and @the-kenny: Starting to look good!! :+1:

Thank you for updating the package!!

**Comment by the-kenny** [comment]
I've got a working `1.10.0` compiler and `0.11.0` cargo locally. I'll test the other packaged rust stuff, maybe bump `rustRegistry` and see if everything works fine.

@Ericson2314 is it okay for you if I open another PR based on my branch? We can continue the discussion there.

**Comment by dvc94ch** [comment]
@the-kenny: I think that's a good idea since you did most of the work.
@Ericson2314: We can take turns criticizing his PR that's more fun anyway =P

**Pull request closed without merging**